### ISSUE TITLE
Example of how to run multiple tasks in parallel?

### ISSUE THREAD

**Issue by sethladd** [google/grinder.dart#170]
We use grinder for dartdoc. We have a few lengthy tasks that can be run in parallel. If any of the tasks fail, we want to fail the build.

Is there an easy way to run multiple tasks in parallel? I mean literally in parallel via isolates.

Thanks!

**Comment by devoncarew** [comment]
Good question. It's not something we've (I've) thought about, but should be possible to do. It'll be easier now that our tasks run in zones - we'll be able to keep the tasks separate. We may want to enable it with additional annotations. It probably needs to be explicit on the part of the user, since a sub-isolates wouldn't share memory with the main one. If the user expected to share state via variables in the script, that wouldn't work if we moved their task into an isolate.

**Comment by sethladd** [comment]
Yup, completely happy to add `@Isolated` or `@Concurrent` to my task.

Would really love this ASAP. Our Travis script is taking a long time, but it doesn't need to :)

**Comment by seaneagan** [comment]
One question is what to do with the output.  If interleaving the output of two concurrent tasks, each line needs to at least be clearly marked with which task it's associated with.  But probably better would be to have one task designated as foreground, and other (background) tasks output is buffered until the task moves to foreground.  Then the question is how to determine which task is foreground.  Maybe a combination of which one is specified first on the command-line, or first in the build file, etc.

Another question is when one task fails do we let other concurrent tasks finish or kill those immediately.

**Comment by seaneagan** [comment]
Currently, based on the tasks specified on the command line and their dependencies, we calculate a task order which is a flat list of tasks.  I think instead we can change that to a list of "phases" which are each a list of tasks which don't have any dependencies on each other, or on tasks in earlier phases.  The order within a phase would be the same as the current task order, whatever that is.  Then we can run all the tasks in a phase concurrently.  We can log a message about which tasks are all starting, and which one is foreground.  Background tasks buffer their output until they become foreground.  As each foreground tasks completes, the next task in the phase becomes foreground, and logs its buffered output.

I think having all tasks support concurrency by default might be the way to go.  Then if we find the need for tasks to share memory, we can add a way to opt out e.g. @NonConcurrent.  Each such task would then probably need to be in its own phase.  We could also potentially add a configuration for max concurrent tasks, like the test runner has, possibly command-line driven.

thoughts?

**Comment by devoncarew** [comment]
A few thoughts:

The above outline makes sense and looks good!

Concurrency will be an issue for scripts not expecting it; The normal dart semantics for globals will not apply (setting up global state in an init task won't apply to tasks that are run in other isolates). We should either have concurrency be opt-in (perhaps on a per-task basis or a per script basis), or have it be opt-out (probably on a per-script basis). So `grind(args, concurrent: false)`.

Perhaps one way to reduce surprise for the script author re concurrency is to print out the execution plan before the build runs? So something like:

```
a ==> b ==> c | d ==> e
```

where the above is 5 tasks and 4 phases.

**Comment by sethladd** [comment]
"shut up and take my money"

**Comment by seaneagan** [comment]
one piece of memory that is shared between tasks is the `context` field.  So we'll need to figure out how to make that available to `spawn`ed isolates.
